userdocs/diagnostics/string-interpolation-conformance.md

```diff
@@ -0,0 +1,12 @@
+# Conforming to `StringInterpolationProtocol`
+A type conforming to `ExpressibleByStringInterpolation` uses a helper type called `StringInterpolation` to perform its interpolation. Many types can use `DefaultStringInterpolation`, which implements `String`'s interpolation behavior. Types can also implement custom behavior by providing their own type conforming to `StringInterpolationProtocol`.
+
+In addition to its formal requirements, `init(literalCapacity:interpolationCount:)` and `appendLiteral(_:)`, `StringInterpolationProtocol` has an additional, informal requirement, `appendInterpolation`. String interpolations using `\()` syntax are translated into calls to matching `appendInterpolation` methods.
+
+`StringInterpolationProtocol` conformers must provide at least one valid `appendInterpolation` method, but it can have any number of overloads. A valid method must:
+
+- Be marked `mutating`
+- Not be marked `static`
+- Either return `Void` or be marked `@discardableResult`
```

Comment:
A requirement not mentioned here: The `appendInterpolation` method's access control must make it at least as visible as the type it's in.

This is a little tricky to explain because less visible `appendInterpolation` methods will still participate in interpolation—there just needs to be at least one that is always visible. I think I might replace the list and introductory paragraph with something like this:

> `StringInterpolationProtocol` conformers must provide at least one `appendInterpolation` method which:
> 
> - Is an instance method (not `static` or `class`)
> - Has no specified return type, explicitly returns `Void`, or returns another type but is marked with `@discardableResult`
> - Has access control which makes it at least as visible as the type it's in
> 
> `appendInterpolation` methods on structs and enums are usually `mutating`, but this isn't required by the compiler. Nor does the compiler place any requirements on an `appendInterpolation` method's argument list, generic-ness, availability, or error-throwing behavior.
>
> If you overload the `appendInterpolation` method, the compiler will choose an appropriate overload for each interpolation. It will consider every `appendInterpolation` instance method accessible at that site, even if they don't meet the other requirements listed above—there just needs to be at least one `appendInterpolation` method which meets all of them.

That should at least give you an idea of the points I would want to touch on—by all means, rewrite it to make it your own, and consider trimming it if you think some of those details are beyond the scope of this educational note.

Reply:
Thanks for the feedback, I've learned a lot about how interpolation works trying to write this 😄 . I've mostly adopted your suggestions, just edited a few parts for length. I also dropped any mention of `mutating`, which is mostly orthogonal and has its own set of error messages.